internal/util/util.go

```diff
@@ -616,3 +617,74 @@ func HttpRequest(method string, url string, headers []string) (resp *http.Respon
 	}
 	return client.Do(req)
 }
+
+// SafeWrite writes bytes to a file in a "safe" way, preventing loss of the
+// contents of the file if it fails to write the new contents.
+// This means that the file is not overwritten directly but by writing to a
+// temporary file first.
+//
+// If rename is true, write is performed atomically, by renaming the temporary
+// file to the target file after the data is successfully written to the
+// temporary file. This guarantees that the file will not remain in a corrupted
+// state, but it also has limitations, e.g. the file should not be a symlink
+// (otherwise SafeWrite silently replaces this symlink with a regular file),
+// the file creation date in Linux is not preserved (since the file inode
+// changes) etc. Use SafeWrite with rename=true for files that are only created
+// and used by micro for its own needs and are not supposed to be used directly
+// by the user.
+//
+// If rename is false, write is performed by overwriting the target file after
+// the data is successfully written to the temporary file.
+// This means that the target file may remain corrupted if overwrite fails,
+// but in such case the temporary file is preserved as a backup so the file
+// can be recovered later. So it is less convenient than atomic write but more
+// universal. Use SafeWrite with rename=false for files that may be managed
+// directly by the user, like settings.json and bindings.json.
+func SafeWrite(path string, bytes []byte, rename bool) error {
+	var err error
+	if _, err = os.Stat(path); err != nil {
+		if !errors.Is(err, fs.ErrNotExist) {
+			return err
+		}
+		return os.WriteFile(path, bytes, FileMode)
```

Comment:
Testing it... When there is no `settings.json` file yet, and I run `set` command to set some option and it creates an empty `settings.json` but fails to write anything to it due to lack of disk space, then micro reports that it failed to write `settings.json` but says nothing about the fact that `settings.json` was actually created but corrupted, and doesn't remove this corrupted file.

So now I'm thinking: to address this corner case as well, it might be better to write newly created files a bit similarly to how we write existing files, namely: first try to create and write to the temporary file (`.micro-backup`), if it succeeds then rename it to the target file (regardless of whether the `rename` argument is true or false), otherwise remove it?

Reply:
Sounds like a good idea.